utils/stringforwarder/stringforwarder_test.go

```diff
@@ -106,27 +99,28 @@ func (*stringForwarderSuite) TestMessagesDroppedWhenBusy(c *gc.C) {
 	c.Check(dropCount, gc.Equals, uint64(3))
 }
 
-func (*stringForwarderSuite) TestRace(c *gc.C) {
-	forwarder := stringforwarder.NewStringForwarder(noopCallback)
+func (*StringForwarderSuite) TestRace(c *gc.C) {
+	forwarder := New(noopCallback)
 	stop := make(chan struct{})
-	go func() {
-		// In 100ms we can hit 1M calls, so make this 10M, and make the
-		// timeout 1ms, that gives us a safety factor of 1000.
-		for i := 0; i < 10*1000*1000; i++ {
+	wg := &sync.WaitGroup{}
+	f := func(wg *sync.WaitGroup) {
+		wg.Done()
+		for {
 			select {
 			case <-stop:
 				return
 			default:
-				forwarder.Receive("next message")
+				forwarder.Forward("next message")
 			}
 		}
-		c.Errorf("managed to send too many messages without being stopped")
-	}()
-	// Note: we intentionally don't use synchronization primatives in the
-	// test suite, because we want to test that the stringForwarder itself
-	// is race free.
-	time.Sleep(1 * time.Millisecond)
-	count := forwarder.Stop()
+	}
+	for i := 0; i < 4; i++ {
+		wg.Add(1)
+		go f(wg)
+	}
+	wg.Wait()
+	time.Sleep(10 * time.Millisecond)
 	close(stop)
+	count := forwarder.Stop()
```

Comment:
Swapping these two lines is the actual fix, right?  While all the clean-up is good, it may have been better to do it separately.  If the patch would have been just moving the one line down then this would have been a trivial ship-it.